File: common/src/main/java/com/ssblur/scriptor/blockentity/LightBlockEntity.java
package com.ssblur.scriptor.blockentity;

import com.ssblur.scriptor.color.interfaces.Colorable;
import net.minecraft.core.BlockPos;
import net.minecraft.core.HolderLookup;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.protocol.Packet;
import net.minecraft.network.protocol.game.ClientGamePacketListener;
import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;
import org.jetbrains.annotations.Nullable;

public class LightBlockEntity extends BlockEntity implements Colorable {
  int color;
  public LightBlockEntity(BlockPos blockPos, BlockState blockState) {
    super(ScriptorBlockEntities.LIGHT.get(), blockPos, blockState);
  }

  public int getColor() {
    return color;
  }

  public void setColor(int color) {
    this.color = color;
    setChanged();
    if(level != null)
      level.sendBlockUpdated(getBlockPos(), getBlockState(), getBlockState(), 2);
  }

  @Nullable
  @Override
  public Packet<ClientGamePacketListener> getUpdatePacket() {
    return ClientboundBlockEntityDataPacket.create(this);
  }

  @Override
  public CompoundTag getUpdateTag(HolderLookup.Provider provider) {
    var tag = super.getUpdateTag(provider);
    tag.putInt("com/ssblur/scriptor/color", color);
    return tag;
  }

  @Override
  public void loadAdditional(CompoundTag tag, HolderLookup.Provider provider) {
    super.loadAdditional(tag, provider);
    color = tag.getInt("com/ssblur/scriptor/color");
    setChanged();
  }

  @Override
  protected void saveAdditional(CompoundTag tag, HolderLookup.Provider provider) {
    super.saveAdditional(tag, provider);

    tag.putInt("com/ssblur/scriptor/color", color);
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/blockentity/LightBlockEntity.kt
package com.ssblur.scriptor.blockentity

import com.ssblur.scriptor.color.interfaces.Colorable
import net.minecraft.core.BlockPos
import net.minecraft.core.HolderLookup
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.protocol.Packet
import net.minecraft.network.protocol.game.ClientGamePacketListener
import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.state.BlockState

class LightBlockEntity(blockPos: BlockPos, blockState: BlockState) :
    BlockEntity(ScriptorBlockEntities.LIGHT.get(), blockPos, blockState), Colorable {
    var colorInternal: Int = 0
    fun getColor(): Int {
        return colorInternal
    }

    override fun setColor(color: Int) {
        this.colorInternal = color
        setChanged()
        if (level != null) level!!.sendBlockUpdated(blockPos, blockState, blockState, 2)
    }

    override fun getUpdatePacket(): Packet<ClientGamePacketListener>? {
        return ClientboundBlockEntityDataPacket.create(this)
    }

    override fun getUpdateTag(provider: HolderLookup.Provider): CompoundTag {
        val tag = super.getUpdateTag(provider)
        tag.putInt("scriptor:color", colorInternal)
        return tag
    }

    public override fun loadAdditional(tag: CompoundTag, provider: HolderLookup.Provider) {
        super.loadAdditional(tag, provider)
        colorInternal = tag.getInt("scriptor:color")
        setChanged()
    }

    override fun saveAdditional(tag: CompoundTag, provider: HolderLookup.Provider) {
        super.saveAdditional(tag, provider)

        tag.putInt("scriptor:color", colorInternal)
    }
}
